### DISCUSSION TITLE
make possible to provide custom implementation of ModelExtension via config

### DISCUSSION THREAD

**Discussion by layerok** [dedoc/scramble#532] [Ideas]
for the context

I use [laravel-translatable](https://spatie.be/docs/laravel-translatable/v6/introduction) library from Spatie.
This library allows you to make model attributes translatable.
You need to add `HasTranslations` trait to your model and specify translatable attributes in $translatable array.

```php
class Product extends Model {
  use HasTranslatable;
  public $translatable = ['title']
}
```
Under the hood `HasTranslations` trait overwrites `getCasts` to automatically assign 'array' cast to all attributes listed in $translatable array.
And also this trait overwrites `getAttributeValue` method. So when you do this
```php
$product->title; // string, not array
```
you get string instead of array.
But Scramble doesn't know this and falsly thinks that `title` attribute is array of strings.

My temporary workaround was to use phpDoc comments in resources

```php
return [
   /** @var string **/
   'title' => $resource->title
]
```
But this is not very convenient.

My current workaround is to overwrite `getPropertyType` method of `ModelExtension`
```php
class ModelExtension implements MethodReturnTypeExtension, PropertyTypeExtension
{
    // ... other mdethods
    public function getPropertyType(PropertyFetchEvent $event): ?Type
    {
        // ... other code
        $info = $this->getModelInfo($event->getInstance());

        if ($attribute = $info->get('attributes')->get($event->getName())) {
            if (in_array(HasTranslations::class, class_uses_recursive($info->get('class')))) {
                $instance = $info->get('instance');
                if (isset($instance->translatable) && in_array($attribute['name'], $instance->translatable)) {
                    return new StringType;
                }
            }
            // ... other code
        }

        // ... other code
    }
    // ... other mdethods
}
```
And everything works fine, but I don't like the way how I overwrote ModelExtension. It seems very fragile to me. There are many implementation details about which I shouldn't know

```php
app()->when(ExtensionsBroker::class)
    ->needs('$extensions')
    ->give(function () {
        $extensions = array_merge(config('scramble.extensions', []), \Dedoc\Scramble\Scramble::$extensions);

        $inferExtensionsClasses = array_values(array_filter(
            $extensions,
            fn ($e) => is_a($e, InferExtension::class, true),
        ));

        $inferExtensionsClasses = array_merge([
            ModelExtension::class, // I changed this
        ], $inferExtensionsClasses);

        return array_merge(
            [
                new PossibleExceptionInfer,
                new AbortHelpersExceptionInfer,

                new JsonResourceCallsTypeInfer,
                new JsonResourceCreationInfer,
                new JsonResourceTypeInfer,
                new ValidatorTypeInfer,
                new ResourceCollectionTypeInfer,
                new ResponseFactoryTypeInfer,
            ],
            array_map(function ($class) {
                return app($class);
            }, $inferExtensionsClasses)
        );
    });
```

It would be really convenient if I could provide my own implementation of ModelExtension via config

```php
return [
     'model_extension' => \Dedoc\Scramble\Support\InferExtensions\ModelExtension::class
]
```

Maybe other infer extensions should also be added to config. For my usecase ModelExtension is enough.

**Comment by layerok** [comment]
And also it would be better if private methods of ModelExtension were protected. Without this it is impossible to reuse methods from base extension

```php
class CustomModelExtension extends ModelExtension {

    public function getPropertyType() {
        $this->getModelInfo(); // Error! Member has private visibility.
    }
} 
```

**Comment by layerok** [comment]
Ok, I found easier way to overwrite ModelExtension

```php
app()->bind(BaseModelExtension::class, CustomModelExtension::class);
```

**Reply by romalytvynenko** [reply]
Exactly!

**Comment by romalytvynenko** [comment]
@layerok have you tried your own custom extension? Like not extending, but using your own

**Comment by martinsjek** [comment]
@romalytvynenko Would it be possible to include this handling into the core?

**Reply by romalytvynenko** [reply]
What do you mean?

**Reply by martinsjek** [reply]
What I am thinking - is this a bug?

Is there any way Scramble can check that translatable fields should return string no array of strings.
Or this is out of this packages Scope, because this translatable functionality comes from an external package.

In the end what I am asking, should we create our own extension for this or could we wait for Scramble.

**Reply by romalytvynenko** [reply]
@martinsjek after taking a quick look, please do create an extension for this. 

This is too specific package behavior.
